File: Next Permutation - GFG/next-permutation.cpp
// { Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
// User function Template for C++

class Solution{
public:
    vector<int> nextPermutation(int n, vector<int> arr){
        if(n<2)
            return arr;
        int lastInc = -1, i = 1;
        while(i<n){
            if(arr[i]>arr[i-1])
                lastInc = i;
            i++;
        }
        if(lastInc==-1){
            for(int i=0;i<n/2;i++)
                swap(arr[i],arr[n-i-1]);
            return arr;
        }
        int index = lastInc;
        for(int i=lastInc;i<n;i++){
            if(arr[i]<arr[index] && arr[i]>arr[lastInc-1])
                index = i;
        }
        swap(arr[lastInc-1],arr[index]);
        sort(arr.begin()+lastInc,arr.end());
        return arr;
    }
};

// { Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int N;
        cin>>N;
        vector<int> arr(N);
        for(int i = 0;i < N;i++)
            cin>>arr[i];
        
        Solution ob;
        vector<int> ans = ob.nextPermutation(N, arr);
        for(int u: ans)
            cout<<u<<" ";
        cout<<"\n";
    }
    return 0;
}  // } Driver Code Ends

File: Next Permutation - GFG/next-permutation.js
// { Driver Code Starts
//Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => {
  inputString += inputStdin;
});

process.stdin.on("end", (_) => {
  inputString = inputString
    .trim()
    .split("\n")
    .map((string) => {
      return string.trim();
    });

  main();
});

function readLine() {
  return inputString[currentLine++];
}

/* Function to print an array */
function printArray(arr, size) {
  let i;
  let s = "";
  for (i = 0; i < size; i++) {
    s += arr[i] + " ";
  }
  console.log(s);
}

function main() {
  let t = parseInt(readLine());
  let i = 0;
  for (; i < t; i++) {
    let n = parseInt(readLine());
    let arr = new Array(n);
    let input_ar1 = readLine().split(" ").map((x) => parseInt(x));
    for (let i = 0; i < n; i++) arr[i] = input_ar1[i];
    let obj = new Solution();
    let res = obj.nextPermutation(arr,n);
    printArray(res,res.length);
  }
}// } Driver Code Ends


//User function Template for javascript

/**
 * @param {Number[]} arr
 * @param {Number} n
 * @returns {Number[]}
 */

class Solution {
    nextPermutation(arr,n){
        if(n<=1)
            return arr;
        let lastInc = -1, i = 1;
        while(i<n){
            if(arr[i]>arr[i-1])
                lastInc = i;
            i++;
        }
        if(lastInc==-1){
            for(let j=0;j<n/2;j++){
                [arr[j],arr[n-j-1]] = [arr[n-j-1],arr[j]];
            }
            return arr;
        }
        let index = lastInc;
        for(let i=lastInc;i<n;i++){
            if(arr[i]<arr[index] && arr[i]>arr[lastInc-1])
                index = i;
        }
        [arr[index],arr[lastInc-1]] = [arr[lastInc-1],arr[index]];
        for(let i=lastInc;i<n-1;i++){
            for(let j=i+1;j<n;j++)
                if(arr[i]>arr[j])
                    [arr[i],arr[j]]=[arr[j],arr[i]];
        }
        return arr;
    }
}